File: backend/app/schemas/slack.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime


class SlackTokensRequest(BaseModel):
    """Request schema for Slack token validation"""
    user_token: str = Field(..., description="Slack user OAuth token (xoxp-)")


class SlackChannel(BaseModel):
    """Slack channel information"""
    id: str = Field(..., description="Channel ID")
    name: str = Field(..., description="Channel name")
    is_private: bool = Field(..., description="Whether channel is private")
    member_count: Optional[int] = Field(None, description="Number of members")
    purpose: Optional[str] = Field(None, description="Channel purpose")
    topic: Optional[str] = Field(None, description="Channel topic")


class SlackChannelsResponse(BaseModel):
    """Response schema for Slack channels"""
    channels: List[SlackChannel]
    team_id: str = Field(..., description="Slack team/workspace ID")
    team_name: str = Field(..., description="Slack team/workspace name")


class SlackConnectionRequest(BaseModel):
    """Request schema for connecting Slack workspace"""
    user_token: str = Field(..., description="Slack user OAuth token")
    selected_channels: List[str] = Field(..., description="List of selected channel IDs", min_items=1, max_items=5)


class SlackConnectionResponse(BaseModel):
    """Response schema for Slack connection"""
    integration_id: str = Field(..., description="Created integration ID")
    team_name: str = Field(..., description="Connected team name")
    channels: List[SlackChannel] = Field(..., description="Connected channels")
    status: str = Field(..., description="Connection status")


class SlackAuthTestResponse(BaseModel):
    """Schema for Slack auth.test API response"""
    ok: bool
    team_id: str
    team: str
    user_id: str
    user: str
    bot_id: Optional[str] = None
    is_enterprise_install: Optional[bool] = None


class SlackConversationsListResponse(BaseModel):
    """Schema for Slack conversations.list API response"""
    ok: bool
    channels: List[dict]
    response_metadata: Optional[dict] = None

File: frontend/src/services/slack.ts
import api from './api';

export interface SlackChannel {
  id: string;
  name: string;
  is_private: boolean;
  member_count?: number;
  purpose?: string;
  topic?: string;
}

export interface SlackTokensRequest {
  user_token: string;
}

export interface SlackChannelsResponse {
  channels: SlackChannel[];
  team_id: string;
  team_name: string;
}

export interface SlackConnectionRequest {
  user_token: string;
  selected_channels: string[];
}

export interface SlackConnectionResponse {
  integration_id: string;
  team_name: string;
  channels: SlackChannel[];
  status: string;
}

export interface SlackIntegration {
  id: string;
  name: string;
  team_name: string;
  team_id: string;
  status: string;
  last_synced?: string;
  channels: Array<{
    id: string;
    name: string;
    is_private: boolean;
    member_count?: number;
  }>;
  created_at: string;
}

class SlackService {
  /**
   * Validate Slack tokens and get available channels
   */
  async validateTokensAndGetChannels(tokens: SlackTokensRequest): Promise<SlackChannelsResponse> {
    const response = await api.post<SlackChannelsResponse>('/api/v1/slack/validate-tokens', tokens);
    return response.data;
  }

  /**
   * Connect Slack workspace with selected channels
   */
  async connectWorkspace(request: SlackConnectionRequest): Promise<SlackConnectionResponse> {
    const response = await api.post<SlackConnectionResponse>('/api/v1/slack/connect', request);
    return response.data;
  }

  /**
   * Get all Slack integrations for the workspace
   */
  async getIntegrations(): Promise<SlackIntegration[]> {
    const response = await api.get<SlackIntegration[]>('/api/v1/slack/integrations');
    return response.data;
  }

  /**
   * Disconnect a Slack integration
   */
  async disconnectIntegration(integrationId: string): Promise<{ message: string }> {
    console.log("slackService.disconnectIntegration called with ID:", integrationId);
    try {
      const response = await api.delete<{ message: string }>(`/api/v1/slack/integrations/${integrationId}`);
      console.log("Disconnect API response:", response.data);
      return response.data;
    } catch (error: any) {
      console.error("Disconnect API error:", error);
      console.error("Error details:", {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status,
        statusText: error.response?.statusText,
        url: error.config?.url,
        method: error.config?.method,
      });
      throw error;
    }
  }
}

export const slackService = new SlackService();